### ISSUE TITLE
ValueError in TextCategorizer.get_loss with multi class classification case

### ISSUE THREAD

**Issue by JulianGerhard21** [explosion/spacy-pytorch-transformers#39]
Python: 3.6.8
Spacy: 2.1.8
Spacy-pytorch-transformers: 0.2.0
Windows 10
Loaded Model: de_pytt_bertbasecased_lg

I am trying to adapt your [sample](https://github.com/explosion/spacy-pytorch-transformers/blob/master/examples/train_textcat.py) to a multi class classification usecase with my own dataset, consisting of 42 classes.

As suggested, I am using:

    textcat = nlp.create_pipe(
        "pytt_textcat", config={"architecture": "softmax_class_vector"}
    )

and afterwards I am manually adding the classes with:

    for label in list(range(0, 42)):
        textcat.add_label(str(label))

Since the sample heads to a binary classification, I had to modify the **_preparte_partition** method like this:

```
def _prepare_partition(text_label_tuples, *, preprocess=False):
    texts, labels = zip(*text_label_tuples)
    if preprocess:
        # Preprocessing can mask errors in our handling of noisy text, so
        # we don't want to do it by default
        texts = [preprocess_text(text) for text in texts]
    cats = [{str(i): 1.0 if i==y else 0.0 for i in range(42)} for y in labels]

    return texts, cats
```

such that I could follow your implementation. If I try to run the finetuning, the following errors occurs:

['sentencizer', 'pytt_wordpiecer', 'pytt_tok2vec']
Loaded model 'de_pytt_bertbasecased_lg'
Loading data...
Using 2557 training docs, 666 evaluation)
Training the model...
LOSS      P       R       F
Traceback (most recent call last):
  File "new_converter.py", line 222, in <module>
    plac.call(main)
  File "bert_spacy_venv\lib\site-packages\plac_core.py", line 328, in call
    cmd, result = parser.consume(arglist)
  File "bert_spacy_venv\lib\site-packages\plac_core.py", line 207, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "new_converter.py", line 95, in main
    nlp.update(texts, annotations, sgd=optimizer, drop=0.1, losses=losses)
  File "bert_spacy_venv\lib\site-packages\spacy_pytorch_transformers\language.py", line 110, in update
    component_cfg=component_cfg,
  File "bert_spacy_venv\lib\site-packages\spacy\language.py", line 475, in update
    proc.update(docs, golds, sgd=get_grads, losses=losses, **kwargs)
  File "pipes.pyx", line 969, in spacy.pipeline.pipes.TextCategorizer.update
  File "bert_spacy_venv\lib\site-packages\spacy_pytorch_transformers\pipeline\textcat.py", line 58, in get_loss
    loss, d_scores = super().get_loss(docs, golds, scores)
  File "pipes.pyx", line 997, in spacy.pipeline.pipes.TextCategorizer.get_loss
ValueError: operands could not be broadcast together with shapes (24,2) (24,42)

I have debugged my way through your implementation and I am really wondering where **(24,2)** might come from. 

Maybe it's just a simple mistake I am making. Any help is appreciated!

**Comment by mkaze** [comment]
I might be wrong, but I think there is a small bug in implementation of `softmax_last_hiddent` architecture in [this line](https://github.com/explosion/spacy-pytorch-transformers/blob/8f2568c29701885255bb5366b2f580d3b3483374/spacy_pytorch_transformers/model_registry.py#L96):

```
return chain(
        get_pytt_last_hidden, flatten_add_lengths, Pooling(mean_pool), Softmax(2, width)
    )
```
The `Softmax(2, width)` enforces the last layer to have 2 neurons, which means it does not respect the number of classes. I think it should be `Softmax(nr_class, width)` instead.

However, you are using `softmax_class_vector` architecture (which does not have the above problem in its implementation). I am not sure why this error happens with that architecture and why there is only 2 neurons in the last layer as shown in the error!

**Comment by JulianGerhard21** [comment]
Just to leave an update: Currently the finetuning process works with using

    textcat = nlp.create_pipe(
        "pytt_textcat", config={"architecture": "softmax_pooler_output"}
    )

as an architecture. As soon as I am trying something else, the mentioned error occurs.

**Comment by honnibal** [comment]
Thanks @mkaze --- fixed by #40
